File: babbleon/events.py
"""
Detection and audit event bus.

The public package emits events; sinks consume them. The default sink
prints to stderr. Enterprise sinks (SIEM forwarder, central escrow alert,
webhook) are registered via the babbleon.event_sinks entry_point group.

Event types are intentionally simple dicts so they cross package boundaries
without import coupling. Each event has at minimum:
  {"event": str, "epoch": int, ...extra fields}
"""

from __future__ import annotations

import logging
import sys
from typing import Any, Protocol

log = logging.getLogger(__name__)


class EventSink(Protocol):
    """Protocol that enterprise event sinks must satisfy."""
    def emit(self, event: dict[str, Any]) -> None: ...


class StderrSink:
    """Default sink: log to stderr. Replace in enterprise with a SIEM forwarder."""
    def emit(self, event: dict[str, Any]) -> None:
        print(f"[babbleon] {event}", file=sys.stderr)


class EventBus:
    """
    Thin multiplexer: emit an event to all registered sinks.

    Enterprise package registers additional sinks at startup:
        bus.add_sink(SIEMForwarder(...))
        bus.add_sink(EscrowAlertSink(...))
    """

    def __init__(self, default_sink: EventSink | None = None) -> None:
        self._sinks: list[EventSink] = [default_sink or StderrSink()]

    def add_sink(self, sink: EventSink) -> None:
        self._sinks.append(sink)

    def emit(self, event: dict[str, Any]) -> None:
        for sink in self._sinks:
            try:
                sink.emit(event)
            except Exception as exc:
                log.warning("event sink %s failed: %s", sink, exc)

    # -- named event helpers --------------------------------------------------

    def honey_triggered(self, *, epoch: int, names: list[str], process_hint: str = "") -> None:
        self.emit({
            "event": "honey_triggered",
            "epoch": epoch,
            "names": names,
            "process_hint": process_hint,
            "severity": "critical",
        })

    def unlock_failed(self, *, epoch: int, backend: str) -> None:
        self.emit({
            "event": "unlock_failed",
            "epoch": epoch,
            "backend": backend,
            "severity": "warning",
        })

    def rotation_complete(self, *, old_epoch: int, new_epoch: int) -> None:
        self.emit({
            "event": "rotation_complete",
            "old_epoch": old_epoch,
            "new_epoch": new_epoch,
            "severity": "info",
        })

    def vault_sealed(self, *, epoch: int, backend: str) -> None:
        self.emit({
            "event": "vault_sealed",
            "epoch": epoch,
            "backend": backend,
            "severity": "info",
        })

File: crates/babbleon/src/events.rs
//! Detection + audit event bus.
//!
//! The public package emits structured events; sinks consume them.
//! Enterprise additions: SIEM forwarder, central escrow alert, webhook.

use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Copy, Serialize, Deserialize)]
pub enum Severity {
    Info,
    Warning,
    Critical,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "event", rename_all = "snake_case")]
pub enum Event {
    HoneyTriggered {
        epoch: u64,
        names: Vec<String>,
        process_hint: String,
    },
    UnlockFailed {
        epoch: u64,
        backend: String,
    },
    RotationComplete {
        old_epoch: u64,
        new_epoch: u64,
    },
    VaultSealed {
        epoch: u64,
        backend: String,
    },
}

impl Event {
    pub fn severity(&self) -> Severity {
        match self {
            Event::HoneyTriggered { .. } => Severity::Critical,
            Event::UnlockFailed { .. } => Severity::Warning,
            _ => Severity::Info,
        }
    }
}

pub trait EventSink: Send + Sync {
    fn emit(&self, event: &Event);
}

pub struct StderrSink;

impl EventSink for StderrSink {
    fn emit(&self, event: &Event) {
        eprintln!("[babbleon] {}", serde_json::to_string(event).unwrap_or_default());
    }
}

/// Append-only JSONL sink for local audit logs.  Each emit takes the
/// write lock briefly; suitable for low-frequency security events.
pub struct JsonlFileSink {
    path: std::path::PathBuf,
    lock: std::sync::Mutex<()>,
}

impl JsonlFileSink {
    pub fn new(path: impl Into<std::path::PathBuf>) -> Self {
        Self {
            path: path.into(),
            lock: std::sync::Mutex::new(()),
        }
    }
}

impl EventSink for JsonlFileSink {
    fn emit(&self, event: &Event) {
        use std::io::Write;
        let _g = self.lock.lock().unwrap_or_else(|e| e.into_inner());
        let line = match serde_json::to_string(event) {
            Ok(s) => s,
            Err(_) => return,
        };
        if let Ok(mut f) = std::fs::OpenOptions::new()
            .create(true)
            .append(true)
            .open(&self.path)
        {
            let _ = writeln!(f, "{line}");
        }
    }
}

pub struct EventBus {
    sinks: Vec<Box<dyn EventSink>>,
}

impl Default for EventBus {
    fn default() -> Self {
        Self {
            sinks: vec![Box::new(StderrSink)],
        }
    }
}

impl EventBus {
    pub fn new() -> Self {
        Self { sinks: Vec::new() }
    }

    pub fn add_sink(&mut self, sink: Box<dyn EventSink>) {
        self.sinks.push(sink);
    }

    pub fn emit(&self, event: Event) {
        for sink in &self.sinks {
            // A panicking sink must not break others.
            let result = std::panic::catch_unwind(std::panic::AssertUnwindSafe(|| sink.emit(&event)));
            let _ = result;
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::sync::{Arc, Mutex};

    struct Capture(Arc<Mutex<Vec<Event>>>);
    impl EventSink for Capture {
        fn emit(&self, e: &Event) {
            self.0.lock().unwrap().push(e.clone());
        }
    }

    #[test]
    fn jsonl_sink_writes_lines() {
        let tmp = tempfile::tempdir().unwrap();
        let path = tmp.path().join("audit.jsonl");
        let sink = JsonlFileSink::new(&path);
        sink.emit(&Event::RotationComplete { old_epoch: 1, new_epoch: 2 });
        sink.emit(&Event::VaultSealed { epoch: 2, backend: "soft".into() });
        let content = std::fs::read_to_string(&path).unwrap();
        let lines: Vec<&str> = content.lines().collect();
        assert_eq!(lines.len(), 2);
        let first: Event = serde_json::from_str(lines[0]).unwrap();
        assert!(matches!(first, Event::RotationComplete { old_epoch: 1, new_epoch: 2 }));
    }

    #[test]
    fn fanout_and_severity() {
        let store = Arc::new(Mutex::new(vec![]));
        let mut bus = EventBus::new();
        bus.add_sink(Box::new(Capture(store.clone())));
        bus.emit(Event::HoneyTriggered {
            epoch: 1,
            names: vec!["x".into()],
            process_hint: "pid=42".into(),
        });
        bus.emit(Event::RotationComplete { old_epoch: 0, new_epoch: 1 });
        let s = store.lock().unwrap();
        assert_eq!(s.len(), 2);
        assert!(matches!(s[0].severity(), Severity::Critical));
        assert!(matches!(s[1].severity(), Severity::Info));
    }
}
